include/boost/numeric/ublas/tensor/functions.hpp

```diff
@@ -24,814 +24,908 @@
-	 * @returns   unary tensor expression
-	*/
-	template<class T, class D>
-	auto real(detail::tensor_expression<T,D> const& expr) {
-		return detail::make_unary_tensor_expression<T> (expr(), [] (auto const& l) { return std::real( l ); } );
-	}
-
-	/** @brief Extract the real component of tensor elements within a tensor expression
-	 *
-	 * @param[in] lhs tensor expression
-	 * @returns   unary tensor expression
-	*/
-	template<class V, class E, class A, class F, class D>
-	auto real(detail::tensor_expression<tensor<std::complex<V>,E,F,A>,D> const& expr)
-	{
-		using tensor_complex_type = tensor<std::complex<V>,E,F,A>;
-		using tensor_type = tensor<V,E,F,typename storage_traits<A>::template rebind<V>>;
-
-		if( detail::retrieve_extents( expr  ).empty() )
-			throw std::runtime_error("error in boost::numeric::ublas::real: tensors should not be empty.");
-
-		auto a = tensor_complex_type( expr );
-		auto c = tensor_type( a.extents() );
-
-		std::transform( a.begin(), a.end(),  c.begin(), [](auto const& l){ return std::real(l) ; }  );
-
-		return c;
-	}
-
-
-	/** @brief Extract the imaginary component of tensor elements within a tensor expression
-	 *
-	 * @param[in] lhs tensor expression
-	 * @returns   unary tensor expression
-	*/
-	template<class T, class D>
-	auto imag(detail::tensor_expression<T,D> const& lhs) {
-		return detail::make_unary_tensor_expression<T> (lhs(), [] (auto const& l) { return std::imag( l ); } );
-	}
-
-
-	/** @brief Extract the imag component of tensor elements within a tensor expression
-	 *
-	 * @param[in] lhs tensor expression
-	 * @returns   unary tensor expression
-	*/
-	template<class V, class E, class A, class F, class D>
-	auto imag(detail::tensor_expression<tensor<std::complex<V>,E,F,A>,D> const& expr)
-	{
-		using tensor_complex_type = tensor<std::complex<V>,E,F,A>;
-		using tensor_type = tensor<V,E,F,typename storage_traits<A>::template rebind<V>>;
-
-		if( detail::retrieve_extents( expr  ).empty() )
-			throw std::runtime_error("error in boost::numeric::ublas::real: tensors should not be empty.");
-
-		auto a = tensor_complex_type( expr );
-		auto c = tensor_type( a.extents() );
-
-		std::transform( a.begin(), a.end(),  c.begin(), [](auto const& l){ return std::imag(l) ; }  );
-
-		return c;
-	}
+    namespace detail{
+        // MSVC 14.1 bug
+        // cannot use ::_size inside constexpr function
+        // basic_extents does not contains _size member because of which
+        // it complains about it
+
+        template<typename E>
+        constexpr auto extents_result_type_tensor_times_vector( E const& ){
+            static_assert(is_static_rank<E>::value, 
+                "boost::numeric::ublas::extents_result_type_tensor_times_vector() : invalid type, type should be an extents");
+            using size_type = typename E::size_type;
+            auto ret = dynamic_extents< std::max( size_type( E::_size - 1 ), size_type(2) )>();
+            ret.fill(typename E::value_type(1));
+            return ret;
```

Comment:
Why is the `extents_result_type_tensor_times_vector` returning a value instead of a type ?

Reply:
Should be also considered for the following code lines.

Reply:
this was because `fixed rank extents` does not have a constructor that fill the extents but it does have `fill` function which is not in `dynamic extents` and I wanted the code clean where I'm calling it.

Reply:
1. I would expect a function named `extents_result_type_tensor_times_vector` to return a `result_type` for `extents` in case of `tensor_times_vector`

2. why does the `tensor_times_vector` function require a helper function like this?

Reply:
It was to convert `static_extents` and `fixed_rank_extents` to convert `fixed_rank_extents ` and `dynamic_extents` to `dynamic_extents`

Reply:
I will change the name to `extents_result_type_tensor_times_vector` to `extents_result_tensor_times_vector `

Reply:
why does the tensor_times_vector function require a helper function like this?

Reply:
It is whether to choose `fixed_rank_extents` or `dynamic_extents` from a given extents type.

Reply:
1. could you use `std::conditional_t` or `std::variant` inside `tensor_times_vector`
2. could you use a free function template for deriving the result types for all kind of contraction operations? 
3. If two is possible this is another small project and we could postpone this

Reply:
I'm not in favor of `std::variant` as it will introduce runtime overhead and the problem with `std::conditional_t ` when you access size, it will complain about a size not being the compile-time constant for dynamic extents. I will think of some kind of solution.